File: step_07_recursion_backtracking/7.3_hard_problems/P002_word_search/Solution.java
/**
 * Problem: Word Search (LeetCode 79)
 * Difficulty: MEDIUM | XP: 25
 *
 * Given an m x n grid of characters and a word, return true if the word
 * exists in the grid. Adjacent means horizontally or vertically neighboring.
 * The same cell may not be used more than once.
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE
    // Time: O(m*n * 4^L)  |  Space: O(L) for visited set + call stack
    //
    // Try starting DFS from every cell. Track visited cells using a
    // HashSet of encoded coordinates to avoid reuse.
    // ============================================================
    public static boolean bruteForce(char[][] board, String word) {
        int rows = board.length, cols = board[0].length;
        Set<Integer> visited = new HashSet<>();

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (dfsBrute(board, word, r, c, 0, visited, rows, cols)) {
                    return true;
                }
            }
        }
        return false;
    }

    private static boolean dfsBrute(char[][] board, String word, int r, int c,
                                     int idx, Set<Integer> visited, int rows, int cols) {
        if (idx == word.length()) return true;
        if (r < 0 || r >= rows || c < 0 || c >= cols) return false;
        int key = r * cols + c;
        if (visited.contains(key)) return false;
        if (board[r][c] != word.charAt(idx)) return false;

        visited.add(key);
        boolean found = dfsBrute(board, word, r + 1, c, idx + 1, visited, rows, cols)
                     || dfsBrute(board, word, r - 1, c, idx + 1, visited, rows, cols)
                     || dfsBrute(board, word, r, c + 1, idx + 1, visited, rows, cols)
                     || dfsBrute(board, word, r, c - 1, idx + 1, visited, rows, cols);
        visited.remove(key);
        return found;
    }

    // ============================================================
    // APPROACH 2: OPTIMAL
    // Time: O(m*n * 4^L)  |  Space: O(L) call stack
    //
    // Same DFS but mark visited cells in-place by replacing char with '#'.
    // Avoids overhead of a HashSet; restore on backtrack.
    // ============================================================
    public static boolean optimal(char[][] board, String word) {
        int rows = board.length, cols = board[0].length;

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (dfsOptimal(board, word, r, c, 0, rows, cols)) {
                    return true;
                }
            }
        }
        return false;
    }

    private static boolean dfsOptimal(char[][] board, String word, int r, int c,
                                       int idx, int rows, int cols) {
        if (idx == word.length()) return true;
        if (r < 0 || r >= rows || c < 0 || c >= cols) return false;
        if (board[r][c] != word.charAt(idx)) return false;

        char temp = board[r][c];
        board[r][c] = '#';

        boolean found = dfsOptimal(board, word, r + 1, c, idx + 1, rows, cols)
                     || dfsOptimal(board, word, r - 1, c, idx + 1, rows, cols)
                     || dfsOptimal(board, word, r, c + 1, idx + 1, rows, cols)
                     || dfsOptimal(board, word, r, c - 1, idx + 1, rows, cols);

        board[r][c] = temp;
        return found;
    }

    // ============================================================
    // APPROACH 3: BEST
    // Time: O(m*n * 4^L)  |  Space: O(L) call stack
    //
    // In-place DFS + frequency pruning:
    // 1. Count characters on board; if word needs more of any char than
    //    available, return false immediately.
    // 2. If word[0] is more frequent on board than word[last], reverse
    //    the word to start searching from the rarer end (fewer starting
    //    points = fewer DFS trees).
    // ============================================================
    public static boolean best(char[][] board, String word) {
        int rows = board.length, cols = board[0].length;

        // Frequency check
        int[] boardFreq = new int[128];
        for (char[] row : board)
            for (char c : row)
                boardFreq[c]++;

        int[] wordFreq = new int[128];
        for (char c : word.toCharArray()) wordFreq[c]++;

        for (int i = 0; i < 128; i++) {
            if (wordFreq[i] > boardFreq[i]) return false;
        }

        // Direction optimization: reverse word if first char is more common
        char[] w = word.toCharArray();
        if (boardFreq[w[0]] > boardFreq[w[w.length - 1]]) {
            // Reverse
            for (int l = 0, r = w.length - 1; l < r; l++, r--) {
                char tmp = w[l]; w[l] = w[r]; w[r] = tmp;
            }
        }

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (dfsBest(board, w, r, c, 0, rows, cols)) {
                    return true;
                }
            }
        }
        return false;
    }

    private static boolean dfsBest(char[][] board, char[] word, int r, int c,
                                    int idx, int rows, int cols) {
        if (idx == word.length) return true;
        if (r < 0 || r >= rows || c < 0 || c >= cols) return false;
        if (board[r][c] != word[idx]) return false;

        char temp = board[r][c];
        board[r][c] = '#';

        boolean found = dfsBest(board, word, r + 1, c, idx + 1, rows, cols)
                     || dfsBest(board, word, r - 1, c, idx + 1, rows, cols)
                     || dfsBest(board, word, r, c + 1, idx + 1, rows, cols)
                     || dfsBest(board, word, r, c - 1, idx + 1, rows, cols);

        board[r][c] = temp;
        return found;
    }

    // ============================================================
    // TESTING
    // ============================================================
    public static void main(String[] args) {
        System.out.println("=== Word Search ===\n");

        Object[][][] tests = {
            {new char[][]{{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}}, "ABCCED", true},
            {new char[][]{{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}}, "SEE",    true},
            {new char[][]{{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}}, "ABCB",   false},
            {new char[][]{{'a'}},                                                 "a",      true},
        };

        for (Object[][] test : tests) {
            char[][] board = (char[][]) test[0];
            String word = (String) test[1];
            boolean expected = (boolean) test[2];

            // Deep copy for each method
            char[][] b1 = copyBoard(board);
            char[][] b2 = copyBoard(board);
            char[][] b3 = copyBoard(board);

            boolean r1 = bruteForce(b1, word);
            boolean r2 = optimal(b2, word);
            boolean r3 = best(b3, word);

            String status = (r1 == expected && r2 == expected && r3 == expected) ? "PASS" : "FAIL";
            System.out.printf("Word: %-10s Expected: %-6b Brute: %-6b Optimal: %-6b Best: %-6b [%s]%n",
                word, expected, r1, r2, r3, status);
        }
    }

    private static char[][] copyBoard(char[][] board) {
        char[][] copy = new char[board.length][];
        for (int i = 0; i < board.length; i++) copy[i] = board[i].clone();
        return copy;
    }
}

File: step_07_recursion_backtracking/7.3_hard_problems/P002_word_search/solution.py
"""
Problem: Word Search (LeetCode 79)
Difficulty: MEDIUM | XP: 25

Given an m x n grid of characters and a word, return True if the word
exists in the grid. The word can be constructed from letters of
sequentially adjacent cells (horizontally or vertically). The same
cell may not be used more than once.
"""
from typing import List
from collections import Counter


# ============================================================
# APPROACH 1: BRUTE FORCE
# Time: O(m*n * 4^L)  |  Space: O(L) recursion stack
#
# Try starting DFS from every cell. At each step try all 4 directions.
# Use a visited set to avoid reusing cells.
# ============================================================
def brute_force(board: List[List[str]], word: str) -> bool:
    """DFS from every cell with a visited set."""
    rows, cols = len(board), len(board[0])
    visited = set()

    def dfs(r: int, c: int, idx: int) -> bool:
        if idx == len(word):
            return True
        if r < 0 or r >= rows or c < 0 or c >= cols:
            return False
        if (r, c) in visited:
            return False
        if board[r][c] != word[idx]:
            return False

        visited.add((r, c))
        found = (dfs(r + 1, c, idx + 1) or
                 dfs(r - 1, c, idx + 1) or
                 dfs(r, c + 1, idx + 1) or
                 dfs(r, c - 1, idx + 1))
        visited.remove((r, c))
        return found

    for r in range(rows):
        for c in range(cols):
            if dfs(r, c, 0):
                return True
    return False


# ============================================================
# APPROACH 2: OPTIMAL
# Time: O(m*n * 4^L)  |  Space: O(L) recursion stack
#
# Same DFS but mark visited cells in-place by replacing char with '#'.
# Avoids the overhead of a separate visited set.
# Early pruning: if board char != word char, don't recurse.
# ============================================================
def optimal(board: List[List[str]], word: str) -> bool:
    """DFS with in-place marking using '#' as visited sentinel."""
    rows, cols = len(board), len(board[0])

    def dfs(r: int, c: int, idx: int) -> bool:
        if idx == len(word):
            return True
        if r < 0 or r >= rows or c < 0 or c >= cols:
            return False
        if board[r][c] != word[idx]:
            return False

        # Mark as visited
        temp = board[r][c]
        board[r][c] = '#'

        found = (dfs(r + 1, c, idx + 1) or
                 dfs(r - 1, c, idx + 1) or
                 dfs(r, c + 1, idx + 1) or
                 dfs(r, c - 1, idx + 1))

        # Restore
        board[r][c] = temp
        return found

    for r in range(rows):
        for c in range(cols):
            if dfs(r, c, 0):
                return True
    return False


# ============================================================
# APPROACH 3: BEST
# Time: O(m*n * 4^L)  |  Space: O(L) recursion stack
#
# Same in-place DFS with two additional optimizations:
# 1. Frequency check: if board doesn't have enough of each letter,
#    return False immediately.
# 2. Direction optimization: if the last character of the word is
#    rarer on the board than the first, reverse the word and search
#    from the rarer end to prune more branches early.
# ============================================================
def best(board: List[List[str]], word: str) -> bool:
    """DFS with in-place marking + frequency pruning + direction optimization."""
    rows, cols = len(board), len(board[0])

    # Frequency check
    board_count = Counter(c for row in board for c in row)
    word_count = Counter(word)
    for ch, cnt in word_count.items():
        if board_count[ch] < cnt:
            return False

    # Direction optimization: start from the rarer end
    if board_count[word[0]] > board_count[word[-1]]:
        word = word[::-1]

    def dfs(r: int, c: int, idx: int) -> bool:
        if idx == len(word):
            return True
        if r < 0 or r >= rows or c < 0 or c >= cols:
            return False
        if board[r][c] != word[idx]:
            return False

        temp = board[r][c]
        board[r][c] = '#'

        found = (dfs(r + 1, c, idx + 1) or
                 dfs(r - 1, c, idx + 1) or
                 dfs(r, c + 1, idx + 1) or
                 dfs(r, c - 1, idx + 1))

        board[r][c] = temp
        return found

    for r in range(rows):
        for c in range(cols):
            if dfs(r, c, 0):
                return True
    return False


# ============================================================
# TESTING
# ============================================================
if __name__ == "__main__":
    print("=== Word Search ===\n")

    tests = [
        (
            [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]],
            "ABCCED", True
        ),
        (
            [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]],
            "SEE", True
        ),
        (
            [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]],
            "ABCB", False
        ),
        (
            [["a"]],
            "a", True
        ),
        (
            [["a","b"],["c","d"]],
            "abdc", True
        ),
    ]

    for board, word, expected in tests:
        import copy
        b1 = copy.deepcopy(board)
        b2 = copy.deepcopy(board)
        b3 = copy.deepcopy(board)
        r1 = brute_force(b1, word)
        r2 = optimal(b2, word)
        r3 = best(b3, word)
        status = "PASS" if r1 == r2 == r3 == expected else "FAIL"
        print(f"Word: {word:<10s} Expected: {str(expected):<6} "
              f"Brute: {str(r1):<6} Optimal: {str(r2):<6} Best: {str(r3):<6}  [{status}]")
